Improve the array task range handling and array concurrency

WilliamMcCumstie opened the pull request:
The range handling has been improved to support `csv` format as well as dash ranges: `1-10` and multiplier ranges: `1-10:3`. The expansion logic has been abstracted into its own class here:
https://github.com/openflighthpc/flight-scheduler-controller/blob/dev/array-ranges/lib/flight_scheduler/range_expander.rb

It also adds the first attempt of concurrency of array tasks. Each "submission" of an "array job" will poll for its available allocated nodes and starts a new task on them. A node is defined as available if it hasn't been assigned to a running Job.

EDIT: The first section of this work has been split out into #13. It will need to be merged first

WilliamMcCumstie committed:
Store the node a task is running on

This will be used to allow array tasks to be ran across multiple
nodes

WilliamMcCumstie committed:
Start a task on all available nodes in array submissions

WilliamMcCumstie committed:
Setup factory bot to create jobs

WilliamMcCumstie committed:
Expand the allocation logic handling

This will be required for latter

WilliamMcCumstie committed:
Create a TaskRegistry for managing array jobs

This will eventually create the array tasks and manage assigning them

WilliamMcCumstie committed:
Allow tasks to transition from pending to running

WilliamMcCumstie committed:
Handle the pending_task being cancelled/finished

This causes it to bypass running_tasks and straight to
the past_tasks

WilliamMcCumstie committed:
Update the running_tasks list as appropriate

WilliamMcCumstie committed:
Add a limit? method to check if using all the nodes

WilliamMcCumstie committed:
Handle when the iteration stops

The pending_task method should start returning nil

WilliamMcCumstie committed:
Start integrating the task registry into the scheduler

WilliamMcCumstie committed:
Treat allocated tasks as running in TaskRegistry

The TaskRegistry needs to treat allocated tasks as if they are running
even though their state has not been updated

WilliamMcCumstie committed:
Correctly skip of jobs which are at their limit

WilliamMcCumstie committed:
Fetch array task's script from the array job

WilliamMcCumstie committed:
Do not clean up after ARRAY_TASKS

This is done by the job

WilliamMcCumstie committed:
Revert the ArrayTask submission handler

WilliamMcCumstie committed:
Remove the obsolete components of message processor

WilliamMcCumstie committed:
Remove the void value expression

Don't know what this is but rspec was complaining

Curiously falcon worked however

WilliamMcCumstie committed:
Add a method for checking if all tasks are finished

WilliamMcCumstie committed:
Remove array jobs in the EventProcessor

This is more idiomatic

WilliamMcCumstie committed:
Run the ARRAY_TASKS in the correct environment

WilliamMcCumstie committed:
Remove Job.array_tasks and create tasks on-demand

Tasks are no longer created up front, instead they are created
when required

WilliamMcCumstie committed:
Start splitting Task from Job

This is only the initial pass to make Job and Task different
objects. There is a lot of conditional duplication between the
two which needs ironing out

WilliamMcCumstie committed:
Serialize the running tasks against the job

WilliamMcCumstie committed:
Update the job state for array tasks

WilliamMcCumstie committed:
Add an aggregate-task to represent all remaining tasks

WilliamMcCumstie committed:
Fix bug where finished tasks where considered running

This was due to a bug in the allocation check

WilliamMcCumstie committed:
Fix bug in determining the last range index

WilliamMcCumstie committed:
Replace aggregate-task with simple integers

This is a bit nicer to deal with client side

benarmston commented on app/models/job.rb:
```diff
@@ -38,20 +38,32 @@ class << self
+      return false unless task_registry.finished?
+      task_registry.past_tasks.reduce(true) do |memo, task|
+        memo && task.complete?
+      end
```
Wouldn't this be better written as `task_registry.past_tasks.all?(&:complete?)`.

WilliamMcCumstie replied:
Indeed: https://github.com/openflighthpc/flight-scheduler-controller/pull/12/commits/58196bee3c34d34d11516bb165f0c465ac016727

benarmston commented on lib/flight_scheduler/range_expander.rb:
```diff
@@ -58,6 +58,15 @@ def valid?
+    end
+
+    # TODO: Make this a private method
     def expand
```
Is there a reason that this hasn't been done?  Does it still need to be done?

WilliamMcCumstie replied:
https://github.com/openflighthpc/flight-scheduler-controller/pull/12/commits/2a88c0fab2126490113ca256fb2bdd207b4335da

benarmston commented on lib/flight_scheduler/schedulers/fifo_scheduler.rb:
```diff
@@ -54,13 +54,14 @@ def add_job(job)
   def remove_job(job)
     @queue.delete(job)
-    job.cleanup
+    job.cleanup unless job.job_type == 'ARRAY_TASK'
```
Is there still a good reason to have `job.cleanup` here instead of in the `EventProcessor`?  It was added here in order to avoid adding it in multiple locations in `EventProcessor`.  But now we're branching on `job_type` to avoid running `job.cleanup` from all the locations that `EventProcessor` calls `FifoScheduler#remove_job`.

WilliamMcCumstie replied:
https://github.com/openflighthpc/flight-scheduler-controller/pull/12/commits/6e54335c39a15c4764133ee3da1a7255f1ae63fa

benarmston commented on lib/flight_scheduler/submission/array_task.rb:
```diff
@@ -27,34 +27,42 @@
 module FlightScheduler::Submission
   class ArrayTask
+    attr_reader :allocation, :job, :task
+
```
What's the reason for making these public?  Previously, the submission was created, `call`ed and it did its thing as a black box.  Are we now doing something with these attributes?

If we're not we should either make these methods private, or, my preferred method, stick to accessing them as instance variables.

WilliamMcCumstie replied:
This is more a stylistic change then anything else. Generally I prefer to use `attr_reader` when defining static attributes within in a constructor. The rubcop style guide is suitable vague on this particular case, the closest I found is this:
https://github.com/rubocop-hq/ruby-style-guide#attr-family

As far as them being public, it doesn't really matter. For the most part I don't bother marking things as private unless I want to make a point.

The bigger problem here is `FlightScheduler::Submission::ArrayTask` violates the [Law of Demeter](https://en.wikipedia.org/wiki/Law_of_Demeter). The accessors are only required because `allocation` is a big complex object which requires unpacking. IMO its the wrong abstraction, however it is out of scope to deal with it now.

tl;dr They changed for stylistic reasons and I've made them private
https://github.com/openflighthpc/flight-scheduler-controller/pull/12/commits/80e232a7462d69290130fdb1250249a721bcc14e

benarmston commented on lib/flight_scheduler/task_registry.rb:
```diff
@@ -0,0 +1,110 @@
+      @running_tasks.select! do |task|
+        (task.running? || (task.allocated? && task.pending?)).tap do |bool|
+          @past_tasks << task unless bool
+        end
```
I think this would be clearer as

```
if task.running? || (task.allocated? && task.pending?)
  @past_tasks << task
end
```

WilliamMcCumstie replied:
Note quite, but point taken: https://github.com/openflighthpc/flight-scheduler-controller/pull/12/commits/fe21552c38c74b7b5cf29551eccbdb09ccf821cb

benarmston commented on lib/flight_scheduler/task_registry.rb:
```diff
@@ -0,0 +1,110 @@
+        rescue StopIteration
+          @pending_task = nil
+        end
+      end
```
I'm not sure what this is trying to do.  Is it the same as this?

```ruby
      # Update the pending task if we have one.
      if @pending_task.nil?
        # Nothing to do.

      elsif !@pending_task.allocated? && @pending_task.pending?
        # Its still pending.  Nothing to do.

      elsif @pending_task.allocated? || @pending_task.running?
        # Its now running.
        @running_tasks << @pending_task
        load_next_pending_task

      else
        # Its completed, or failed, or cancelled.
        @past_tasks << @pending_task
        load_next_pending_task
      end
```

WilliamMcCumstie replied:
Not quite. The `allocated?` stuff was a bit of a curve ball from the scheduler. It tries to allocate the resources before starting any tasks and thus does not update the `state` field. This was preventing the `TaskRegistry` from progressing any further. The fix is to make the `TaskRegistry` aware that allocated tasks are as good as "running".

This has introduced a couple of misnomer, mainly `running_tasks` may contain tasks which are in a pseudo `PENDING`-"allocated" state. The `pending_task` method has been renamed `next_task` to try and clear up the confusion between the two.

There have been a couple of commits cleaning this section up. The main one being:
https://github.com/openflighthpc/flight-scheduler-controller/pull/12/commits/013287eac0b2c31559894e317402d86ae2078553

benarmston commented on lib/flight_scheduler/task_registry.rb:
```diff
@@ -0,0 +1,110 @@
+
+  def limit?(update = true)
+    refresh if update
+    @running_tasks.length >= job.min_nodes
```
This really ought to be testing against `max_nodes` not `min_nodes`.  However, we don't have that at the moment.  I think it will be worth adding that soon and updating this method at that point.

WilliamMcCumstie replied:
Yeah the`max_nodes` :eyes: 
https://github.com/openflighthpc/flight-scheduler-controller/pull/12/commits/67263e067c0be3bf57543d7ae877c4e7058b1df7

benarmston commented on spec/task_registry_spec.rb:
```diff
@@ -0,0 +1,202 @@
+      end
+
+      it 'returns nil after all tasks have been started' do
+        (max + 2).times { subject.pending_task&.state = 'RUNNING' }
```
Why `+ 2`?  The job has 10 tasks, but we need to set 12 to have a state of `RUNNING` before pending task is `nil`?  That doesn't seem right.  Is something wrong here or am I having a Monday?

WilliamMcCumstie replied:
Ah you are right, this was a hangover from an older implementation
https://github.com/openflighthpc/flight-scheduler-controller/pull/12/commits/f895a49352dc949ce68219f781ad6b180a6de780

benarmston commented on spec/task_registry_spec.rb:
```diff
@@ -0,0 +1,202 @@
+    end
+
+    describe '#limit?' do
+      it { should_not be_limit }
```
Reading this spec I'm not convinced that `limit?` is named correctly.  `limit?` returns true if the maximum permitted number of running tasks are running.  Perhaps `max_tasks_running?`?

WilliamMcCumstie replied:
https://github.com/openflighthpc/flight-scheduler-controller/pull/12/commits/5755583c921a08b4b2038da73d69c914f87fbd48

benarmston reviewed (changes requested):
The Job/Task factoring work you've done seems like a good fit for array
jobs/tasks, but I don't think its going to be correct going forwards.

I've come to the following conclusions whilst trying to get interactive
sessions working in a manner that I'm happy with.

1. Jobs are things to which resources are allocated.  If resources need to be
	 allocated, they must be allocated to a job.

2. A Job may or may not have some kind of "payload" associated with it.  We
	 need to be able to allocate resources without having anything to do with
	 those resources.  We also need to be able to allocate resources with a
	 known "payload".

3. We need to be able add "payloads" to an allocated/running job whilst it is
	 running.

Let's keep the Job/Task factoring you've done for now.  But don't be surprised
to see much of this change this week when interactive sessions land.

---



I get the following error occasionally.

```
ben@glitnir 14:24:05 flight-scheduler (dev/list-tasks %=)
$ ./bin/scheduler batch --array 1-7:2 --nodes 2 tmp/simple-array.sh 
Submitted batch job d68abf11-dd7a-4001-a35d-ad85dba6414e
ben@glitnir 14:24:10 flight-scheduler (dev/list-tasks %=)
$ ./bin/scheduler queue
┌───────────────────────────────────────────┬───────────┬─────────────────┬──────┬─────────┬──────┬────────┬──────────────────┐
│ JOBID                                     │ PARTITION │ NAME            │ USER │ ST      │ TIME │ NODES  │ NODELIST(REASON) │
├───────────────────────────────────────────┼───────────┼─────────────────┼──────┼─────────┼──────┼────────┼──────────────────┤
│ d68abf11-dd7a-4001-a35d-ad85dba6414e[5-7] │ all       │ simple-array.sh │ TBD  │ PENDING │ TBD  │ (none) │ (Resources)      │
│ d68abf11-dd7a-4001-a35d-ad85dba6414e[1]   │ all       │ simple-array.sh │ TBD  │ RUNNING │ TBD  │ 1      │ node01           │
│ d68abf11-dd7a-4001-a35d-ad85dba6414e[3]   │ all       │ simple-array.sh │ TBD  │ RUNNING │ TBD  │ 1      │ node02           │
└───────────────────────────────────────────┴───────────┴─────────────────┴──────┴─────────┴──────┴────────┴──────────────────┘
ben@glitnir 14:24:11 flight-scheduler (dev/list-tasks %=)
$ ./bin/scheduler queue
bin/scheduler: The API returned a 500 error status and this content:
  {
    "errors": [
      {
        "id": "d7acb939-d316-46bd-a08a-af6a33f74354",
        "title": "No Method Error",
        "detail": "undefined method `array_index' for nil:NilClass",
        "status": "500"
      }
    ]
  }
```



```
28m23s    debug: Falcon::Server [oid=0xff0] [pid=23805] [2020-09-28 14:24:22 +0100]
               | Incoming connnection from #<Addrinfo: 127.0.0.1:42754 TCP> to Async::HTTP::Protocol::HTTP1
2020-09-28 14:24:22 - NoMethodError - undefined method `array_index' for nil:NilClass:
	/home/ben/projects/flight-scheduler-controller/app/serializers.rb:71:in `block in <class:JobSerializer>'

	<SNIP>

E, [2020-09-28T14:24:22.569761 #23805] ERROR -- sinja: {:id=>"974587f1-b5ed-4f49-a1cf-f1cd248eeca7", :title=>"No Method Error", :detail=>"undefined method `array_index' for nil:NilClass", :status=>"500"}
127.0.0.1 - - [28/Sep/2020:14:24:22 +0100] "GET /v0/jobs?include=partition%2Callocated-nodes%2Crunning-tasks%2Crunning-tasks.allocated-nodes HTTP/1.1" 500 208 0.0020
```

WilliamMcCumstie committed:
Use the all? helper instead of reduce

WilliamMcCumstie committed:
Make the accessors private methods

WilliamMcCumstie committed:
Simplify the logic when updating @past_tasks

WilliamMcCumstie committed:
Add a guard to filter out nil @pending_task

WilliamMcCumstie committed:
Rename pending_task to next_task

WilliamMcCumstie committed:
Simplify the nested logic within the TaskRegistry

WilliamMcCumstie committed:
Attempt to break the iteration by randomly overrunning

WilliamMcCumstie committed:
Remove the random addition in the nil check

This was a fix to a hangover in an older spec. It should only run
the `max` number of times

WilliamMcCumstie committed:
Rename limit? to max_tasks_running?

WilliamMcCumstie committed:
Use the "max_nodes" to determine the task limit

WilliamMcCumstie committed:
Make the `expand` method private

The public cached `expanded` method should be used instead

WilliamMcCumstie committed:
Preform the appropriate cleanup in EventProcessor

This separates the concerns from the Scheduler

WilliamMcCumstie committed:
Fix bug where `next_task` maybe nil

WilliamMcCumstie commented:
The undefined `array_index` bug has been fixed here:
https://github.com/openflighthpc/flight-scheduler-controller/pull/12/commits/f9935db170c8da404dae6999d3d82194d3b1c356

benarmston commented on app/models/job.rb:
```diff
@@ -38,20 +38,30 @@ class << self
     define_method("#{s.downcase}?") { self.state == s }
   end
+
+  # TODO: Check if the complexity in the complete? method is required
```
Why would this not be required?  What would the alternative be?  If we can simplify this by setting the state correctly in event processor, let's do so as part of this PR.

WilliamMcCumstie replied:
Looking through the code base the `complete?` method isn't called anywhere. Overriding the `complete?` method but not `failed?`, `cancelled?`, etc.. looks a bit strange IMO.

Whilst we have the opportunity I have removed the override: https://github.com/openflighthpc/flight-scheduler-controller/pull/12/commits/87ac9ae0e46968fc7f855449e54fdd5e3d3901ad
Instead the `complete?` method is a short hand to `job.state == 'COMPLETED'`

Whilst I agree setting the `state` could use a simplification any solution would require architectural changes. This is best left for a future PR.

benarmston replied:
As discussed, can we update `task.array_job` in `node_failed_task` and `node_completed_task` too.

benarmston commented on app/models/job.rb:
```diff
@@ -38,20 +38,30 @@ class << self
+  # The node an individual task is ran on. This is not used by the "job" variety of Job
+  # NOTE: Do not expose this publicly! It is solely used by the scheduler to track
+  #       which task is running where
+  attr_accessor :task_node
```
I don't think that this is used anywhere.  Perhaps its left over from a previous iteration.  If it is still used, but only by `Task`s, it would be better to define it on `Task`.

WilliamMcCumstie replied:
Good spot, removed: https://github.com/openflighthpc/flight-scheduler-controller/pull/12/commits/fd44c472fb81b692adc82cc796ab05ce06682bfd

benarmston commented on app/models/job.rb:
```diff
@@ -182,13 +179,40 @@ def hash
+end
+
+class Task
+  # TODO: Reinstate these validations
+  # Validations for `JOB`s and `ARRAY_JOB`s.
+  # validates :array_index,
+  #   absence: true, unless: ->() { job_type == 'ARRAY_TASK' }
+  # validates :array_job,
+  #   absence: true, unless: ->() { job_type == 'ARRAY_TASK' }
+  # Validations for `ARRAY_TASK`s
+  # validates :array_index,
+  #   presence: true,
+  #   numericality: { allow_blank: false, only_integer: true, greater_than_or_equal_to: 0 },
+  #   if: ->() { job_type == 'ARRAY_TASK' }
```
What's needed to reinstate these?  If its quick let's do it.

WilliamMcCumstie replied:
I've taken a different approach and remove the comment block entirely: https://github.com/openflighthpc/flight-scheduler-controller/pull/12/commits/226f421ff338abfc631d960fda4358557112ef1a

The `Task` objects should not be invalid as `Job` is validated up front and `TaskRegistry` is tested extensively. This means validating the `Task` gives no advantages.

benarmston replied:
That's fine.

benarmston commented on lib/flight_scheduler/task_registry.rb:
```diff
@@ -0,0 +1,123 @@
+
+  def refresh
+    @mutex.synchronize do
+      @running_tasks.select! do |task|
+        # Transition "finalised" tasks from running to past
+        if task.running? || (task.allocated? && task.pending?)
+          true
+        else
+          @past_tasks << task
+          false
+        end
+      end
```
This is easier to understand than it was before.  However, I still have some issues with it.

This method is modifying mutable instance variables within the `synchronize` block.  If say, `running_tasks(false)` or `past_tasks(false)` are called whilst this method is running in some other thread, they will get the partially updated `@running_tasks`/`@past_tasks`.  This could result in a task appearing twice or not appearing at all.

We can reduce that race condition like this (totally untested code follows):

```ruby
@mutex.synchronize do
  now_running, now_past = @running_tasks.partition { |task| task.running? || (task.allocated? && task.pending?) }
  # Handle next task as before, but using now_running and now_past instead of @running_tasks and @past_tasks
  @running_tasks = now_running
  @past_tasks.concat(now_past)
end
```

Replacing `select!` with `partition` is also easier to understand its purpose.

WilliamMcCumstie replied:
Done: https://github.com/openflighthpc/flight-scheduler-controller/pull/12/commits/1266e37367c51b8d31741a803c0293acb2a8ebf0

WilliamMcCumstie replied:
And a bug fix + spec: https://github.com/openflighthpc/flight-scheduler-controller/pull/12/commits/1d3b990168e1243edec95b3c3b6e5b14128e9b55

benarmston replied:
This is fine for now.

benarmston reviewed (changes requested):
This looks much better than it did.  I've left a few more comments I'd like you to look at before I merge this.

WilliamMcCumstie committed:
Fix bug where deleting a job didn't cleanup the script

WilliamMcCumstie committed:
Remove the complete? override method

It isn't used and I'm not convinced it is correct

WilliamMcCumstie committed:
Remove the obsolete task_node method

WilliamMcCumstie committed:
Remove the validations on the Task object

WilliamMcCumstie committed:
Use partition instead of select!

WilliamMcCumstie committed:
Fix bug where past tasks where forgotten

benarmston commented on app.rb:
```diff
@@ -266,6 +266,7 @@ def validate!
     end
 
     destroy do
+      resource.cleanup
```
Can we move this to `EventProcessor.cancel_job`.  We're eventually going to want to handle job cancellations differently, for the different types of jobs.  Having all of the logic in a single location is going to make that easier to refactor.

WilliamMcCumstie replied:
Done in: https://github.com/openflighthpc/flight-scheduler-controller/pull/12/commits/3e18100d8075d5ed369cc957614f64a33dee2612

benarmston replied:
Thanks.

WilliamMcCumstie committed:
Protect all task_registry methods with a mutex

WilliamMcCumstie committed:
Update the array jobs final state

WilliamMcCumstie committed:
Cleanup the job within EventProcessor.cancel_job

WilliamMcCumstie commented:
Made a couple of final changes as discussed IRL
